File: core/date_range.rb
require 'date'

class DateRange
  attr_reader :start_date, :end_date

  def initialize(start_date, end_date)
    raise ArgumentError, "La date de fin doit être postérieure ou égale à la date de début" if end_date < start_date

    @start_date = start_date
    @end_date = end_date
  end

  def collides?(other)
    (start_date <= other.start_date && other.start_date <= end_date) ||
    (start_date <= other.end_date && other.end_date <= end_date)
  end

  def to_s
    "#{start_date} à #{end_date}"
  end
end

File: core/date_range.py
from datetime import date

class DateRange:
    def __init__(self, start: date, end: date):
        if end < start:
            raise ValueError("La date de fin doit être postérieure ou égale à la date de début")
        self.start = start
        self.end = end

    def collides(self, other: 'DateRange') -> bool:
        return (
            (self.start <= other.start <= self.end) or
            (self.start <= other.end <= self.end)
        )

    def __str__(self):
        return f"{self.start.isoformat()} à {self.end.isoformat()}"
